Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -24,30 +31,142 @@ public class DafnyHoverHandler : HoverHandlerBase {
       };
     }
 
-    public async override Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
+    public override async Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
       logger.LogTrace("received hover request for {Document}", request.TextDocument);
       var document = await documents.GetDocumentAsync(request.TextDocument);
       if (document == null) {
         logger.LogWarning("the document {Document} is not loaded", request.TextDocument);
         return null;
       }
+      var diagnosticHoverContent = GetDiagnosticsHover(document, request.Position, out var areMethodStatistics);
       if (!document.SymbolTable.TryGetSymbolAt(request.Position, out var symbol)) {
         logger.LogDebug("no symbol was found at {Position} in {Document}", request.Position, request.TextDocument);
+      }
+
+      var symbolHoverContent = symbol != null ? CreateSymbolMarkdown(symbol, cancellationToken) : null;
+      if (diagnosticHoverContent == null && symbolHoverContent == null) {
         return null;
       }
-      return CreateHover(symbol, cancellationToken);
+
+      // If diagnostics are method diagnostics, we prioritize displaying the symbol information.
+      // This makes testing easier and less surprise for the user.
+      var hoverContent = areMethodStatistics && symbolHoverContent != null ? "" : (diagnosticHoverContent ?? "");
+      hoverContent = symbolHoverContent != null ? hoverContent + (hoverContent != "" ? "  \n" : "") + symbolHoverContent : hoverContent;
+      return CreateMarkdownHover(hoverContent);
+    }
+
+    private string? GetDiagnosticsHover(DafnyDocument document, Position position, out bool areMethodStatistics) {
+      areMethodStatistics = false;
+      foreach (var node in document.VerificationTree.Children.OfType<TopLevelDeclMemberVerificationTree>()) {
+        if (node.Range.Contains(position)) {
+          var assertionBatchCount = node.AssertionBatchCount;
+          var assertionBatchIndex = -1;
+          var information = "";
+          foreach (var assertionBatch in node.AssertionBatches) {
+            assertionBatchIndex += 1;
```

Comment:
It might not always be the case that the assertion batches are listed in order. This should probably use `VCResult.vcNum`, piped through appropriately. The assertions in each batch should always be in the same order, though.

Reply:
I did not know about  `vcNum`, that's a great insight ! I refactored the code to make use of it instead of a counter that is incrementing. It will even make the migration of assertion batch diagnostics easier down the road since this counter is probably more deterministic than mine.